src/storage.py

```diff
@@ -0,0 +1,283 @@
+            directive = storage.split("/")[0]
+            storage_path = self.juju_storage_get(storage_id=storage, attribute="location")
+            if not self._get_osd_num(storage_path, directive):
+                enroll[directive].append(storage_path)
+
+        # enrolls available disks with DB and save osd data.
+        with sunbeam_guard.guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_with_db(disk=enroll[self.dedicated_db], db=enroll["db"])
+            self.charm.status.set(ActiveStatus("charm is ready"))
+
+    def _on_osd_standalone_attached(self, event: StorageAttachedEvent):
+        """Storage attached handler for osd-standalone."""
+        if not self.charm.ready_for_service():
+            logger.warning("MicroCeph not ready yet, deferring storage event.")
+            event.defer()
+            return
+
+        self._clean_stale_osd_data()
+
+        enroll = []
+        for storage in self._fetch_filtered_storages([self.standalone]):
+            path = self.juju_storage_get(storage_id=storage, attribute="location")
+            if not self._get_osd_num(path, self.standalone):
+                enroll.append(path)
+
+        with sunbeam_guard.guard(self.charm, self.name):
+            self.charm.status.set(MaintenanceStatus("Enrolling OSDs"))
+            self._enroll_disks_in_batch(enroll)
+            self.charm.status.set(ActiveStatus("charm is ready"))
+
+    def _on_storage_detaching(self, event: StorageDetachingEvent):
+        """Unified storage detaching handler."""
+        # check if the detaching device is being used as or with an OSD.
+        osd_num = self._get_osd_num(event.storage.location.as_posix(), event.storage.name)
+
+        if osd_num:
+            with sunbeam_guard.guard(self.charm, self.name):
+                try:
+                    self.remove_osd(osd_num)
+                except CalledProcessError as e:
+                    if self._is_safety_failure(e.stderr):
+                        warning = f"Storage {event.storage.full_id} detached, provide replacement for osd.{osd_num}."
+                        logger.warning(warning)
+                        # clean records since juju will deprovision device.
+                        self.remove_osd(osd_num, force=True)
+                        raise sunbeam_guard.BlockedExceptionError(warning)
+
+    # helper functions
+
+    def _fetch_filtered_storages(self, directives: list) -> list:
+        """Provides a filtered list of attached storage devices."""
+        filtered = []
+        for device in self.juju_storage_list():
+            if device.split("/")[0] in directives:
+                filtered.append(device)
+
+        return filtered
+
+    def _is_safety_failure(self, err: str) -> bool:
+        """Checks if the subprocess error is caused by safety check."""
+        return "need at least 3 OSDs" in err
```

Comment:
There could be other faults too though, eg. if ceph signals it's unsafe to stop the OSD

Reply:
It may be more useful to check the `errno` return than playing games with stdout strings, which are much more volatile

Reply:
So, there is not much uniqueness in-terms of error_num returned by go errors. Unfortunately we are left with string comparisons to make certain of errors. We may extend the error interface to also map to a unique error num (this however lies outside the scope of this PR).
```
utkarsh@workbook:~$ sudo microceph disk add /dev/vdz --encrypt

+----------+---------+
|   PATH   | STATUS  |
+----------+---------+
| /dev/vdz | Failure |
+----------+---------+
Error: failed to set stable disk path: invalid disk path: /dev/vdz
utkarsh@workbook:~$ echo $?
1
utkarsh@workbook:~$ sudo microceph cluster config add
Error: unknown command "add" for "microceph cluster config"
utkarsh@workbook:~$ echo $?
1
utkarsh@workbook:~$ sudo microceph cluster config get "my_network"
Error: Key my_network is invalid. 
Permitted Keys: [cluster_network osd_pool_default_crush_rule]
utkarsh@workbook:~$ echo $?
1
```